test: Add test for wtxid transaction relay

fjahr opened the pull request:
Based on #18044 

This updates the functional testing framework to the p2p protocol version including wtxid transaction relay (70016) and adds a basic test for wtxid based relay.

sdaftuar committed:
Add a wtxid-index to the mempool

sdaftuar committed:
Add a wtxid-index to mapRelay

sdaftuar committed:
Add wtxid-index to orphan map

sdaftuar committed:
Add wtxids of confirmed transactions to bloom filter

This is in preparation for wtxid-based invs (we need to be able to tell whether
we AlreadyHave() a transaction based on either txid or wtxid).

This also double the size of the bloom filter, which is overkill, but still
uses a manageable amount of memory.

sdaftuar committed:
Add wtxids to recentRejects instead of txids

Previously, we only added txids to recentRejects if we were sure that the
transaction couldn't have had the wrong witness (either because the witness was
malleated or stripped).

In preparation for wtxid-based relay, we can observe that txid == wtxid for
transactions that have no witness, and add the wtxid of rejected transactions,
provided the transaction wasn't a witness-stripped one. This means that we now
add more data to the filter (as prior to this commit, any transaction with a
witness that failed to be accepted was being skipped for inclusion in the
filter) but witness malleation should still not interfere with relay of a valid
segwit transaction, because the txid of a segwit transaction would not be added
to the filter after failing validation.

In the future, having wtxids in the recent rejects filter will allow us to
skip downloading the same wtxid multiple times, once our peers use wtxids for
transaction relay.

sdaftuar committed:
Add support for tx-relay via wtxid

This adds a field to CNodeState that tracks whether to relay transactions with
that peer via wtxid, instead of txid. As of this commit the field will always
be false, but in a later commit we will add a way to negotiate turning this on
via p2p messages exchanged with the peer.

sdaftuar committed:
Add p2p message "wtxidrelay"

When sent to and received from a given peer, enables using wtxid's for
announcing and fetching transactions with that peer.

sdaftuar committed:
Just pass a hash to AddInventoryKnown

Since it's only used for transactions, there's no need to pass in an inv type.

sdaftuar committed:
Delay getdata requests from peers using txid-based relay

Using both txid and wtxid-based relay with peers means that we could sometimes
download the same transaction twice, if announced via two different hashes from
different peers.

Use a heuristic of delaying txid-peer-getdata requests by 2 seconds, if we have
at least one wtxid-based peer.

sdaftuar committed:
Make TX_WITNESS_STRIPPED its own rejection reason

Previously, TX_WITNESS_MUTATED could be returned during transaction validation
for either transactions that had a witness that was non-standard, or for
transactions that had no witness but were invalid due to segwit validation
rules.

However, for txid/wtxid-relay considerations, net_processing distinguishes the
witness stripped case separately, because it affects whether a wtxid should be
able to be added to the reject filter. It is safe to add the wtxid of a
witness-mutated transaction to the filter (as that wtxid shouldn't collide with
the txid, and hence it wouldn't interfere with transaction relay from
txid-relay peers), but it is not safe to add the wtxid (== txid) of a
witness-stripped transaction to the filter, because that would interfere with
relay of another transaction with the same txid (but different wtxid) when
relaying from txid-relay peers.

Also updates the comment explaining this logic, and explaining that we can get
rid of this complexity once there's a sufficient deployment of wtxid-relaying
peers on the network.

sdaftuar committed:
Rename AddInventoryKnown() to AddKnownTx()

fjahr committed:
test: Update test framework p2p protocol version to 70016

This new p2p protocol version allows to use WTXIDs for tx relay.

fjahr committed:
test: Add tests for wtxid tx relay in segwit test

Also cleans up some doublicate lines in the rest of the test.

co-authored-by: Anthony Towns <aj@erisian.com.au>

sipa commented on test/functional/test_framework/messages.py:
```diff
@@ -31,7 +31,7 @@
 
 MIN_VERSION_SUPPORTED = 60001
-MY_VERSION = 70014  # past bip-31 for ping/pong
+MY_VERSION = 70016  # past bip-31 for ping/pong
```
Comment is outdated now.

fjahr replied:
fixed

ajtowns commented:
I think this isn't testing:

 * whether the node switches from sending TX invs to WTX invs when the wtxidrelay message is sent
 * whether we get a wtxidrelay message when we give a 70016 version, but not when we give a lower version
 * in either case, when we announce a TX inv, do we get a TX getdata back, and likewise for WTX?
 * when a tx is sent (not an INV but the actual tx) if it's an orphan, the missing parent's txid is requested as a TX not a WTX

I've had a go at some of the changes: https://github.com/ajtowns/bitcoin/tree/202003-wtxidrelay-test but haven't made it so it actually sends the `wtxidrelay` message.

fjahr commented:
@ajtowns Thank you, that is very helpful. My intuition was that many of these cases would be tested indirectly by upgrading the test framework but of course, explicit tests are always better and, as I have realized now, I also had not implemented it correctly. Your tests mainly did not pass because I had not implemented feature negotiation correctly. This is done now and I preserved your commit mostly so it is easier for you to review. If the test now does what you intended it to do then I can squash.

Overview of changes:
- Rebased
- Fixed outdated comment
- Added explicit test for wtxidrelay
- Fixed wtxidrelay feature negotiation
- Added @ajtowns code and with some further changes on top

ajtowns reviewed (changes requested):
I think this needs more changes to match the latest update to #18044; I merged them and added a fixup commit on https://github.com/ajtowns/bitcoin/tree/202003-wtxidrelay-test if you want to take a look. Feel free to squash the fixups anytime btw.

fjahr commented on test/functional/p2p_segwit.py:
```diff
@@ -143,16 +146,24 @@ def test_witness_block(node, p2p, block, accepted, with_witness=True, reason=Non
+        if self.wtxidrelay:
+            self.send_message(msg_wtxidrelay())
+        self.send_message(msg_verack())
+        self.nServices = message.nServices
```
done

fjahr commented:
Thanks @ajtowns ! I pulled in latest changes in #18044 and your suggestions. Then squashed and edited commit messages for easier reviewing. Also left your suggested change `ignore non-wtxidrelay compliant invs` in for now, pending discussion in #18044 .

fjahr committed:
test: Use wtxid relay generally in functional tests

fjahr commented:
Pushed an update that should make all functional tests use wtxid relay by default (Github doesn't seem to be working properly at the moment, here is the commit: https://github.com/fjahr/bitcoin/commit/becdb68ab37fdcbc1dcf743b19ea8629f52677bb)

amitiuttarwar committed:
Add wtxid to mempool unbroadcast tracking

sdaftuar commented:
FYI -- I tried to cherry-pick the test commits here onto the latest version of #18044 but the p2p_segwit.py test is failing for me.

fjahr commented:
> FYI -- I tried to cherry-pick the test commits here onto the latest version of #18044 but the p2p_segwit.py test is failing for me.

@sdaftuar Thanks for the heads-up! I rebased with the latest state of #18044 and made several fixes plus squashed to three commits now.

fjahr commented:
These commits were pulled into #18044 so it can be closed.
